src/Util/ClassSourceManipulator.php

```diff
@@ -0,0 +1,992 @@
+            $quoteValue = !(is_bool($value) || is_int($value) || is_array($value));
+
+            if (is_bool($value)) {
+                $value = $value ? 'true' : 'false';
+            }
+
+            if (is_array($value)) {
+                if (!isset($value[0])) {
+                    // associative array: we'll add this if/when we need it
+                    throw new \Exception('Not currently supported');
+                }
+
+                $value = sprintf('{%s}', implode(', ', array_map(function ($val) {
+                    return sprintf('"%s"', $val);
+                }, $value)));
+            }
+
+            if ($quoteValue) {
+                return sprintf('%s="%s"', $option, $value);
+            }
+
+            return sprintf('%s=%s', $option, $value);
+        }, array_keys($options), array_values($options));
+
+        return sprintf('%s(%s)', $annotationClass, implode(', ', $formattedOptions));
+    }
+
+    private function addSingularRelation(BaseRelation $relation)
+    {
+        $typeHint = $this->addUseStatementIfNecessary($relation->getTargetClassName());
+        if ($relation->getTargetClassName() == $this->getThisFullClassName()) {
+            $typeHint = 'self';
+        }
+
+        $annotationOptions = [
+            'targetEntity' => $relation->getTargetClassName(),
+        ];
+        if ($relation->isOwning()) {
+            // sometimes, we don't map the inverse relation
+            if ($relation->getMapInverseRelation()) {
+                $annotationOptions['inversedBy'] = $relation->getTargetPropertyName();
+            }
+        } else {
+            $annotationOptions['mappedBy'] = $relation->getTargetPropertyName();
+        }
+
+        if ($relation instanceof RelationOneToOne) {
+            $annotationOptions['cascade'] = ['persist', 'remove'];
+        }
+
+        $annotations = [
+            $this->buildAnnotationLine(
+                $relation instanceof RelationManyToOne ? '@ORM\\ManyToOne' : '@ORM\\OneToOne',
+                $annotationOptions
+            ),
+        ];
+
+        if (!$relation->isNullable() && $relation->isOwning()) {
+            $annotations[] = $this->buildAnnotationLine('@ORM\\JoinColumn', [
+                'nullable' => false,
+            ]);
+        }
+        $this->addProperty($relation->getPropertyName(), $annotations);
+
+        $this->addGetter(
+            $relation->getPropertyName(),
+            $typeHint,
+            // getter methods always have nullable return values
+            // because even though these are required in the db, they may not be set
+            true
+        );
+
+        $setterNodeBuilder = $this->createSetterNodeBuilder(
+            $relation->getPropertyName(),
+            $typeHint,
+            // make the type-hint nullable always for ManyToOne to allow the owning
+            // side to be set to null, which is needed for orphanRemoval
+            // e.g. $userAvatarPhoto->setUser(null);
```

Comment:
that's wrong. OrphanRemoval does not require setting it to `null`. It only requires removing it from the inverse collection.

Reply:
or, the case of ManyToOne vs OnetoOne are swapped

Reply:
Yea, this is confusing. The issue is that the generated `remove*` method on the inverse side (and really, the mutators for all the inverse sides) synchronizes the owning side - e.g. it looks like this:

```php
// User.php
public function removeAvatarPhoto(UserAvatarPhoto $photo)
{
    $this->avatarPhotos->removeElement($photo);
    // This is the key: we sync the owning side, which is why it needs to be nullable
    $photo->setUser(null);
}
```

When orphanRemoval is set, yea, we don't need to sync the owning side, as the orphanRemoval will cause the deletion. But since we *are* sync'ing it (and I think that's generally a good idea - and I don't want to NOT sync it just for the case of `orphanRemoval=true` - want to not create so many variations), we need to allow the setter to be null. 

These are the tough balances

Reply:
but this syncing forces you to make the setter accept null on the owning side, even if `null` would be an invalid value because of the field being non-nullable, allowing invalid usage of the class.